tests/integration/dynamic_ip_test.py

```diff
@@ -671,72 +661,146 @@ def _get_nameservers():
     """
     Return a list of name server string configured in RESOLV_CONF_PATH.
     """
-    return (
+    running_ns = (
         libnmstate.show()
         .get(Constants.DNS, {})
         .get(DNS.RUNNING, {})
         .get(DNS.SERVER, [])
     )
+    logging.debug("Current running DNS: {}".format(running_ns))
+    return running_ns
 
 
 def _get_running_routes():
     """
     return a list of running routes
     """
-    return libnmstate.show().get(Constants.ROUTES, {}).get(RT.RUNNING, [])
-
-
-def _has_ipv6_auto_gateway():
-    routes = _get_running_routes()
-    for route in routes:
-        if (
-            route[RT.DESTINATION] == IPV6_DEFAULT_GATEWAY
-            and route[RT.NEXT_HOP_INTERFACE] == DHCP_CLI_NIC
-        ):
-            return True
-    return False
-
-
-def _has_ipv6_auto_extra_route():
-    routes = _get_running_routes()
-    for route in routes:
-        if (
-            route[RT.DESTINATION] == IPV6_CLASSLESS_ROUTE_DST_NET1
-            and route[RT.NEXT_HOP_INTERFACE] == DHCP_CLI_NIC
-        ):
-            return True
-    return False
-
-
-def _has_ipv6_auto_nameserver():
-    return DHCP_SRV_IP6 in _get_nameservers()
-
-
-def _has_ipv4_dhcp_nameserver():
-    return DHCP_SRV_IP4 in _get_nameservers()
-
-
-def _has_ipv4_dhcp_gateway():
-    routes = _get_running_routes()
-    for route in routes:
-        if (
-            route[RT.DESTINATION] == IPV4_DEFAULT_GATEWAY
-            and route[RT.NEXT_HOP_INTERFACE] == DHCP_CLI_NIC
-        ):
-            return True
-    return False
-
-
-def _has_ipv4_classless_route():
-    routes = _get_running_routes()
-    for route in routes:
-        if (
-            route[RT.DESTINATION] == IPV4_CLASSLESS_ROUTE_DST_NET1
-            and route[RT.NEXT_HOP_ADDRESS] == IPV4_CLASSLESS_ROUTE_NEXT_HOP1
-            and route[RT.NEXT_HOP_INTERFACE] == DHCP_CLI_NIC
-        ):
-            return True
-    return False
+    running_routes = (
+        libnmstate.show().get(Constants.ROUTES, {}).get(RT.RUNNING, [])
+    )
+    logging.debug("Current running routes: {}".format(running_routes))
+    return running_routes
+
+
+def _poll(func, timeout):
+    ret = func()
+    while timeout > 0:
+        if ret:
+            break
+        time.sleep(1)
+        timeout -= 1
+        ret = func()
+    return ret
+
+
+def _has_ipv6_auto_gateway(poll=True, nic=DHCP_CLI_NIC):
+    def _check():
+        routes = _get_running_routes()
+        for route in routes:
+            if (
+                route[RT.DESTINATION] == IPV6_DEFAULT_GATEWAY
+                and route[RT.NEXT_HOP_INTERFACE] == nic
+            ):
+                return True
+        return False
+
+    return _poll(_check, DEFAULT_TMO if poll else 0)
```

Comment:
IMHO instead of moving `_poll()` into the `has_*` functions, the caller should call `_poll()` explicitly, for example:

```
assert _poll(_has_ipv6_auto_nameserver)
```

It would be less code and nicer to read and understand. Not sure if `_poll` is the best name.
```